src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/RSACryptoServiceProvider.cs

```diff
@@ -64,29 +65,99 @@ private RSACryptoServiceProvider(int keySize, CspParameters parameters, bool use
             // in the constructor so we can report any errors now.
             if (!_randomKeyContainer)
             {
-                GetKeyPair();
+                // Force-read the SafeKeyHandle property, which will summon it into existence.
+                SafeKeyHandle localHandle = SafeKeyHandle;
+                Debug.Assert(localHandle != null);
             }
         }
 
-        /// <summary>
-        /// Retreives the key pair
-        /// </summary>
-        private void GetKeyPair()
+        private SafeProvHandle SafeProvHandle
         {
-            if (_safeKeyHandle == null)
+            get
+            {
+                if (_safeProvHandle == null)
+                {
+                    lock (_parameters)
+                    {
+                        if (_safeProvHandle == null)
+                        {
+                            SafeProvHandle hProv = CapiHelper.CreateProvHandle(_parameters, _randomKeyContainer);
+
+                            Debug.Assert(hProv != null);
+                            Debug.Assert(!hProv.IsInvalid);
+                            Debug.Assert(!hProv.IsClosed);
+
+                            _safeProvHandle = hProv;
+                        }
+                    }
+
+                    return _safeProvHandle;
+                }
+
+                return _safeProvHandle;
+            }
+            set
             {
-                lock (this)
+                lock (_parameters)
                 {
-                    if (_safeKeyHandle == null)
+                    if (ReferenceEquals(value, _safeProvHandle))
                     {
-                        // We only attempt to generate a random key on desktop runtimes because the CoreCLR
-                        // RSA surface area is limited to simply verifying signatures.  Since generating a
-                        // random key to verify signatures will always lead to failure (unless we happend to
-                        // win the lottery and randomly generate the signing key ...), there is no need
-                        // to add this functionality to CoreCLR at this point.
-                        CapiHelper.GetKeyPairHelper(CapiHelper.CspAlgorithmType.Rsa, _parameters,
-                                            _randomKeyContainer, _keySize, ref _safeProvHandle, ref _safeKeyHandle);
+                        return;
                     }
+
+                    if (_safeProvHandle != null)
+                    {
+                        _safeKeyHandle?.Dispose();
+                        _safeKeyHandle = null;
+
+                        _safeProvHandle?.Dispose();
```

Comment:
We just checked that _safeProvHandle isn't null... we don't need to check again.  This can just be _safeProvHandle.Dispose().